### DISCUSSION TITLE
How to use middleware for a specific route only in a single APIRouter

### DISCUSSION THREAD

**Discussion by ahnaf-zamil** [fastapi/fastapi#6494] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
async def auth_required_middleware(request: Request, call_next: Coroutine):
    """Middleware for authenticating users on required routes"""
    token = str(
        request.cookies.get("token")
    )  # Stringyfy-ing because pyjwt might cry if it gets NoneType

    try:
        user_id = JWTService.verify_and_return_id(token)
        request.state.user_id = user_id
        response: StreamingResponse = await call_next(request)
        return response
    except:
        return JSONResponse(content={"description": "Unauthorized"}, status_code=401)
```

### Description

FastAPI has a feature to have sub-apps. But it's not applicable in all cases, especially in my case where I have an `APIRouter` instance with 3 path operations/routes: `login_user`, `create_user`, `get_logged_in_user`. And I have an authentication middlware.

I spent a lot of time trying to find a way to only use that middleware in the `get_logged_in_user` operation, and not the other ones. But I found nothing regarding this. All I found online was using sub-apps, but I was trying to find a way to be able to use middlewares for a few specific routes in the same `APIRouter` instance.

So is there some kind of way where I can attach a middleware to a select few routes/operations in a single APIRouter instance? Because at the moment, the middleware is being triggered on all routes, including my user create and login path operations; which isn't practical.

This is how I'm adding my middleware, since it's in a separate file.
```py
app.middleware("http")(
    auth_required_middleware
)  # Using decorators the way they were not meant to be ig
```

### Operating System

Linux

### Operating System Details

Linux Mint. Kernel version: 5.4.0-74-generic

DE: Not using one, just DWM

### FastAPI Version

0.72.0

### Python Version

Python 3.8.10

### Additional Context

I have just taken a look at the source code, and it seems that the FastAPI middleware system is actually handled by Starlette. So
does that mean per-route/per-APIRouter middleware is something unrelated to FastAPI but related to Starlette itself?

**Comment by STeveShary** [comment, marked as answer]
@ahnaf-zamil Rather than think about this as a middleware, you could add in a single line a reusable function to get the logged in user.  An example to look at would be here: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-in-path-operation-decorators/

Another pattern using the middleware is to define a list of "unauthenticated" endpoints.   And you only have the middleware apply the user checking if the path does not match any of the unauthenticated endpoints.

**Comment by ahnaf-zamil** [comment]
Ah thanks a lot! I will be closing this issue then

**Comment by WilliamStam** [comment]
ive been scratching my head at this for the last few days. i must be doing something totaly out there with my other non py projects :( 

consider an api with "groups" like (as an example)
```
/api 
    /instances
       /{instance}
           /update
           /info
           /accounts
               /{account}
    /system
    /profile
        /update 
```

i would normally attach a middleware to `/api/instances/*` to lookup the instance in `{instance}` and pass the instance object to all child route groups and endpoints., the same for accounts

as far as i can see the Depends option wont allow for passing on any values to the routes only raising errors or doing arbetary stuff that isnt linked to the end pages "life?'


in other systems i use middleware to for instance in the case of `/api/{instance}/{account}`

-----

there is the option of each middleware group actualy being a fastapi app itself (which is what im going to try next), but then do you "loose" your higher up the foodchain middleware then? unless you explicitly include it for each sub app(group).

**Reply by ditsuke** [reply]
Hi @WilliamStam, I'm attempting a similar pattern here in 2024. Could you share what you learnt from your experiments?

**Reply by WilliamStam** [reply]
sure. ignore "middleware". its a trap. think of middleware more as "application code" than route code then you're fine. 

use dependencies. 

```
async def get_instance(
    instance_id: int = Path(),
) -> model.InstanceModel:
    instance = await repository.get(instance_id)
    if instance.id is None:
        raise HTTPNotFoundException(204, f"Instance with id {instance_id} not found")
    return instance


Instance = typing.Annotated[model.InstanceModel, Depends(get_instance)]
```

then in your route 

```
@router.get("/{instance_id}", description="Retrieve an Instance using Id as unique Key")
async def get_an_instance(
    instance: Instance,
) -> InstanceResponse:
    response = InstanceResponse()
    if instance:
        response.instance = instance

    return response
```

you can keep using this `instance: Instance` wherever it must lookup the path param,eter of "instance_id" and return an object. 

```
@router.get("/{instance_id}/{account_nr}")
async def get_an_instance(
    instance: Instance,
    account: Account
)
@router.get("/{instance_id}/{account_nr}/{statemement_id}")
async def get_an_instance(
    instance: Instance,
    account: Account,
    statement: Statement
)
```

if you dont need instance and account in the routebut you do need it to lookup the statement 

```
async def get_statement(
    instance_id: int = Path(),
    account_nr: str = Path(),
    statemement_id: int = Path()
) -> model.StatementModel:
    result = await repository.get(instance_id, account=account_nr, statement=statement_id)
    if result .id is None:
        raise HTTPNotFoundException(204, f"statement with id {statement_id} not found")
    return result 


Statement= typing.Annotated[model.StatementModel, Depends(get_statement)]

# your dependencies can lookup other dependencies whenever they run.
@router.get("/{instance_id}/{account_nr}/{statemement_id}")
async def get_an_instance(
    statement: Statement
)

```

the responses of those dependencies are cached at least so if you call instance: Instance multiple times on a route as in the route handler and in other dependencies it will only look it up once so no performance hit there

--------
you cant unfortunately do something with this "after" the route has ran. 

as a bonus tip for the "after". we want to log every page request, since you cant use middleware to lookup the auth methods so nicely we use a dependency attached to the app level. that then accepts `request: Request` and adds the user to the scope there. and then in a middleware that runs after the route we read from request. would defs have prefered being able to do it all in 1 middleware but meh.

**Reply by ditsuke** [reply]
Thanks, that is really creative use of dependencies and I never quite imagined dependencies can do a lookup for other dependencies! My initial idea was to use a custom decorator for routes -- this is so much cleaner when the use-case ends before the handler call.
